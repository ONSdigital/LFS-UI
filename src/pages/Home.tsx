import React, {ChangeEvent, Component} from "react";
import {HomeBatchTable} from "../components/HomeBatchTable";
import {ONSRadioButton} from "../components/ONS_DesignSystem/ONSRadioButton";
import {ONSCheckbox} from "../components/ONS_DesignSystem/ONSCheckbox";
import {getAllBatches} from "../utilities/http";
import DocumentTitle from "react-document-title";
import {ONSButton} from "../components/ONS_DesignSystem/ONSButton";
import {Link} from "react-router-dom";

const MONTHLY_BATCH = "Monthly";
const QUARTERLY_BATCH = "Quarterly";
const ANNUALLY_BATCH = "Annually";

interface State {
    batchData: [] | null,
    filteredBatchData: [] | null,
    batchType: string,
    liveStatus: boolean,
    monthlyBatchFilter: boolean,
    quarterlyBatchFilter: boolean,
    annuallyBatchFilter: boolean
}

export class Home extends Component <{}, State> {
    displayName = Home.name;

    constructor(props: any) {
        super(props);
        this.state = {
            batchData: [],
            filteredBatchData: [],
            batchType: "live",
            liveStatus: true,
            monthlyBatchFilter: true,
            quarterlyBatchFilter: true,
            annuallyBatchFilter: true
        };
        this.getBatchData();
    }

    getBatchData = () => {
        getAllBatches()
            .then(r => {
                console.log(r);
                this.setState({batchData: r});
                this.filterBatchData();
            })
            .catch(error => {
                console.log(error);
                if (process.env.NODE_ENV === "development") {
                    console.log("Error Connecting to Server, Loading mock data");
                    this.getMockBatchData();
                }
            });
    };

    getMockBatchData = () => {
        fetch("/jsons/MOCK_RUNS.json")
            .then(response => response.json())
            .then(response => {
                this.setState({batchData: response.Rows});
                this.filterBatchData();
            });
    };

    filterListByStatus = (row: any): boolean => {
        if (this.state.batchType === "completed") {
            return (row.status === 4);
        }
        return (row.status !== 4);
    };

    filterListByBatchType = (row: any): boolean => {
        let monthly = false;
        let quarterly = false;
        let annually = false;
        if (this.state.monthlyBatchFilter) {
            monthly = (row.type === MONTHLY_BATCH);
        }
        if (this.state.quarterlyBatchFilter) {
            quarterly = (row.type === QUARTERLY_BATCH);
        }
        if (this.state.annuallyBatchFilter) {
            annually = (row.type === ANNUALLY_BATCH);
        }
        return monthly || quarterly || annually;
    };

    filterBatchData = () => {
        if (this.state.batchData !== null) {
            let filteredList: any = this.state.batchData.filter(this.filterListByBatchType);
            filteredList = filteredList.filter(this.filterListByStatus);
            if (filteredList !== null) {
                this.setState({filteredBatchData: filteredList});
            }
        }
    };

    handleBatchTypeRadioChange = async (e: ChangeEvent<HTMLInputElement>) => {
        let liveStatusFilter = false;
        if (e.target.value === "live") {
            liveStatusFilter = true;
        }
        await this.setState({batchType: e.target.value, liveStatus: liveStatusFilter});
        this.filterBatchData();
    };

    handleMonthlyBatchFilterChange = async () => {
        await this.setState({monthlyBatchFilter: !this.state.monthlyBatchFilter});
        this.filterBatchData();
    };

    handleQuarterlyBatchFilterChange = async () => {
        await this.setState({quarterlyBatchFilter: !this.state.quarterlyBatchFilter});
        this.filterBatchData();
    };

    handleAnnuallyBatchFilterChange = async () => {
        await this.setState({annuallyBatchFilter: !this.state.annuallyBatchFilter});
        this.filterBatchData();
    };

    filterOptionStyle = {marginRight: ".5rem", minWidth: "15rem"};

    render() {
        return (
            <DocumentTitle title={"Labour Force Survey Dashboard"}>
                <div className="container">
                    <br/>
                    <fieldset className="fieldset">
                        <Link to={"/new-batch"} style={{right: 0, float: "right"}}>
                            <ONSButton label={"Create New Batch"} primary={true} field={true} action={true}/>
                        </Link>
                        <h4 className="fieldset__legend">Filter Batches</h4>
                        <p className="checkboxes__label">Batch Status</p>
                        <span className="radios__items">
                        <ONSRadioButton label={"Live"}
                                        onChange={this.handleBatchTypeRadioChange}
                                        id={"live"}
                                        checked={this.state.liveStatus}
                                        style={this.filterOptionStyle}/>
                        <ONSRadioButton label={"Completed"}
                                        onChange={this.handleBatchTypeRadioChange}
                                        id={"completed"}
                                        checked={!this.state.liveStatus}
                                        style={this.filterOptionStyle}/>
                            <br/>
                    </span>
                        <p className="checkboxes__label">Batch Type</p>
                        <span className="checkboxes__items">
                        <ONSCheckbox onChange={this.handleMonthlyBatchFilterChange}
                                     id={MONTHLY_BATCH}
                                     label={MONTHLY_BATCH}
                                     checked={this.state.monthlyBatchFilter}
                                     style={this.filterOptionStyle}/>
                        <ONSCheckbox onChange={this.handleQuarterlyBatchFilterChange}
                                     id={QUARTERLY_BATCH}
                                     label={QUARTERLY_BATCH}
                                     checked={this.state.quarterlyBatchFilter}
                                     style={this.filterOptionStyle}/>
                        <ONSCheckbox onChange={this.handleAnnuallyBatchFilterChange}
                                     id={ANNUALLY_BATCH}
                                     label={ANNUALLY_BATCH}
                                     checked={this.state.annuallyBatchFilter}
                                     style={this.filterOptionStyle}/>
                    </span>
                    </fieldset>
                    <HomeBatchTable data={this.state.filteredBatchData}/>
                </div>
            </DocumentTitle>
        );
    }
}

