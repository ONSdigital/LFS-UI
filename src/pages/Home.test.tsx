import React from "react";

import {default as flushPromises} from "../tests/util/flushPromises";
import {cleanup, fireEvent, render} from "@testing-library/react";

import {BrowserRouter} from "react-router-dom";

import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import fetch from "../tests/setup/__mocks__/fetch.js";
import WS from "jest-websocket-mock";
import {Home} from "./Home";


// @ts-ignore
global.fetch = fetch;


describe("Home Page", () => {
    Enzyme.configure({adapter: new Adapter()});

    let server: WS;

    // beforeEach(() => {
    //     server = new WS("ws://127.0.0.1:8000/ws", {jsonProtocol: true});
    // });

    afterEach(() => {
        cleanup();
    });

    const Props = {};


    function wrapper(render: any, props: any) {
        return render(
            <BrowserRouter>
                <Home/>
            </BrowserRouter>
        );
    }

    // /*
    // Function to Mock a file in the File Selection, then press the Submit button
    //  */
    // async function selectAndImportFile(getByTestId: any, getByLabelText: any, fileName: string) {
    //     const inputEl = getByLabelText(/import/i);
    //
    //     const file = new File(["(⌐□_□)"], fileName, {
    //         type: "sav"
    //     });
    //
    //     Object.defineProperty(inputEl, "files", {
    //         value: [file]
    //     });
    //
    //     fireEvent.change(inputEl);
    //
    //     fireEvent.click(getByTestId("import-button"));
    //
    //     await act(async () => {
    //         await flushPromises();
    //     });
    // }

    it("renders the page", async () => {
        expect(wrapper(render, Props)).toMatchSnapshot();
    });

    it("should display all open batch items in the list", async () => {
        const {getByTestId, getByLabelText, getByText, queryAllByTestId} = wrapper(render, Props);

        await flushPromises();

        // Check the
        let list = queryAllByTestId(/table-row/i);

        let listItemOne = list[0];
        let firstRowData = listItemOne.children[2].textContent + " " + listItemOne.children[3].textContent + " " + listItemOne.children[4].textContent;
        expect(firstRowData).toEqual("Monthly April 2019");

        let listItemTwo = list[5];
        let secondRowData = listItemTwo.children[2].textContent + " " + listItemTwo.children[3].textContent + " " + listItemTwo.children[4].textContent;
        expect(secondRowData).toEqual("Monthly May 2019");

        let listItemThree = list[10];
        let thirdRowData = listItemThree.children[2].textContent + " " + listItemThree.children[3].textContent + " " + listItemThree.children[4].textContent;
        expect(thirdRowData).toEqual("Monthly June 2019");

        let listItemFour = list[15];
        let fourthRowData = listItemFour.children[2].textContent + " " + listItemFour.children[3].textContent + " " + listItemFour.children[4].textContent;
        expect(fourthRowData).toEqual("Quarterly Q2 2019");
    });

    it("should filter the Batch list by its status", async () => {
        const {getByTestId, getByLabelText, getByText, getAllByText, queryAllByTestId} = wrapper(render, Props);

        await flushPromises();

        // Filter by Completed
        await fireEvent.click(getAllByText(/Completed/i)[0]);

        await flushPromises();

        // Check the
        let tableList = queryAllByTestId(/table-row/i);

        let listItemOne = tableList[0];
        let firstRowData = listItemOne.children[2].textContent + " " + listItemOne.children[3].textContent + " " + listItemOne.children[4].textContent;
        expect(firstRowData).toEqual("Monthly January 2019");

        let listItemTwo = tableList[5];
        let secondRowData = listItemTwo.children[2].textContent + " " + listItemTwo.children[3].textContent + " " + listItemTwo.children[4].textContent;
        expect(secondRowData).toEqual("Monthly February 2019");

        let listItemThree = tableList[10];
        let thirdRowData = listItemThree.children[2].textContent + " " + listItemThree.children[3].textContent + " " + listItemThree.children[4].textContent;
        expect(thirdRowData).toEqual("Monthly March 2019");

        let listItemFour = tableList[15];
        let fourthRowData = listItemFour.children[2].textContent + " " + listItemFour.children[3].textContent + " " + listItemFour.children[4].textContent;
        expect(fourthRowData).toEqual("Quarterly Q1 2019");

        // Filter by Live again
        await fireEvent.click(getByText(/Live/i));

        await flushPromises();

        let list = queryAllByTestId(/table-row/i);
        listItemOne = list[0];
        firstRowData = listItemOne.children[2].textContent + " " + listItemOne.children[3].textContent + " " + listItemOne.children[4].textContent;
        console.log(firstRowData);
        expect(firstRowData).toEqual("Monthly April 2019");
    });

    it("should filter the Batch list by Batch Type", async () => {
        const {getByTestId, getByLabelText, getByText, getAllByText, queryAllByTestId} = wrapper(render, Props);

        await flushPromises();

        // Uncheck the monthly batch type checkbox
        await fireEvent.click(getAllByText(/Monthly/i)[0]);

        await flushPromises();

        let list = queryAllByTestId(/table-row/i);

        let listItemOne = list[0];
        let firstRowData = listItemOne.children[2].textContent + " " + listItemOne.children[3].textContent + " " + listItemOne.children[4].textContent;
        expect(firstRowData).toEqual("Quarterly Q2 2019");

        let listItemTwo = list[5];
        let secondRowData = listItemTwo.children[2].textContent + " " + listItemTwo.children[3].textContent + " " + listItemTwo.children[4].textContent;
        expect(secondRowData).toEqual("Annually 2019 2019");


        // Uncheck the Quarterly batch type checkbox
        await fireEvent.click(getAllByText(/Quarterly/i)[0]);

        await flushPromises();

        list = queryAllByTestId(/table-row/i);
        listItemOne = list[0];
        firstRowData = listItemOne.children[2].textContent + " " + listItemOne.children[3].textContent + " " + listItemOne.children[4].textContent;
        expect(firstRowData).toEqual("Annually 2019 2019");

        // Uncheck the Annually batch type checkbox
        await fireEvent.click(getAllByText(/Annually/i)[0]);

        await flushPromises();

        expect(getByText(/No Batches matching this criteria/i)).toBeDefined()

        console.log(process.env.NODE_ENV)
    });

    it("should load test data in development environment", async () => {
        // @ts-ignore - Set Test Env to development
        process.env.NODE_ENV = "development";
        const {getByTestId, getByLabelText, getByText, getAllByText, queryAllByTestId} = wrapper(render, Props);

        await flushPromises();

    });




});