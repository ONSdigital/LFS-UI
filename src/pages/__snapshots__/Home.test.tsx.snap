// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page renders the page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <br />
        <fieldset
          class="fieldset"
        >
          <a
            href="/new-batch"
            style="right: 0px; float: right;"
          >
            <button
              class="btn btn--link field "
              data-testid="button"
              type="button"
            >
              <span
                class="btn__inner"
              >
                Create New Batch
              </span>
            </button>
          </a>
          <h4
            class="fieldset__legend"
          >
            Filter Batches
          </h4>
          <p
            class="checkboxes__label"
          >
            Batch Status
          </p>
          <span
            class="radios__items"
          >
            <p
              class="radios__item"
              style="margin-right: .5rem; min-width: 15rem;"
            >
              <span
                class="radio"
              >
                <input
                  checked=""
                  class="radio__input js-radio"
                  id="live"
                  name="Live"
                  type="radio"
                  value="live"
                />
                <label
                  class="radio__label "
                  for="live"
                >
                  Live
                </label>
              </span>
            </p>
            <p
              class="radios__item"
              style="margin-right: .5rem; min-width: 15rem;"
            >
              <span
                class="radio"
              >
                <input
                  class="radio__input js-radio"
                  id="completed"
                  name="Completed"
                  type="radio"
                  value="completed"
                />
                <label
                  class="radio__label "
                  for="completed"
                >
                  Completed
                </label>
              </span>
            </p>
            <br />
          </span>
          <p
            class="checkboxes__label"
          >
            Batch Type
          </p>
          <span
            class="checkboxes__items"
          >
            <p
              class="checkboxes__item"
              style="margin-right: .5rem; min-width: 15rem;"
            >
              <span
                class="checkbox "
              >
                <input
                  checked=""
                  class="checkbox__input js-checkbox "
                  id="Monthly"
                  name="Monthly"
                  type="checkbox"
                  value="Monthly"
                />
                <label
                  class="checkbox__label "
                  for="Monthly"
                >
                  Monthly
                </label>
              </span>
            </p>
            <p
              class="checkboxes__item"
              style="margin-right: .5rem; min-width: 15rem;"
            >
              <span
                class="checkbox "
              >
                <input
                  checked=""
                  class="checkbox__input js-checkbox "
                  id="Quarterly"
                  name="Quarterly"
                  type="checkbox"
                  value="Quarterly"
                />
                <label
                  class="checkbox__label "
                  for="Quarterly"
                >
                  Quarterly
                </label>
              </span>
            </p>
            <p
              class="checkboxes__item"
              style="margin-right: .5rem; min-width: 15rem;"
            >
              <span
                class="checkbox "
              >
                <input
                  checked=""
                  class="checkbox__input js-checkbox "
                  id="Annually"
                  name="Annually"
                  type="checkbox"
                  value="Annually"
                />
                <label
                  class="checkbox__label "
                  for="Annually"
                >
                  Annually
                </label>
              </span>
            </p>
          </span>
        </fieldset>
        <table
          class="table table--sortable"
          data-aria-asc="ascending"
          data-aria-desc="descending"
          data-aria-sort="Sort by"
          id="basic-table"
        >
          <thead
            class="table__head"
          >
            <tr
              class="table__row"
            >
              <th
                aria-sort="none"
                class="table__header "
                scope="col"
              />
              <th
                class="table__header "
                scope="col"
              >
                BatchID
              </th>
              <th
                class="table__header "
                scope="col"
              >
                Batch Type
              </th>
              <th
                aria-sort="none"
                class="table__header "
                scope="col"
              >
                <button
                  aria-label="Sort by Period"
                  class="table__sort-button"
                  data-index="2"
                  type="button"
                >
                  Period
                </button>
              </th>
              <th
                aria-sort="none"
                class="table__header "
                scope="col"
              >
                <button
                  aria-label="Sort by Year"
                  class="table__sort-button"
                  data-index="3"
                  type="button"
                >
                  Year
                </button>
              </th>
              <th
                aria-sort="none"
                class="table__header "
                scope="col"
              >
                <button
                  aria-label="Sort by Status"
                  class="table__sort-button"
                  data-index="4"
                  type="button"
                >
                  Status
                </button>
              </th>
            </tr>
          </thead>
          <tbody
            class="table__body expandedRowEnabled "
          >
            <tr>
              <td
                class="table__cell "
                colspan="6"
              >
                <div
                  class="panel panel--info panel--simple "
                >
                  <div
                    class="panel__body"
                  >
                    <p>
                      No Batches matching this criteria
                    </p>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <br />
      <fieldset
        class="fieldset"
      >
        <a
          href="/new-batch"
          style="right: 0px; float: right;"
        >
          <button
            class="btn btn--link field "
            data-testid="button"
            type="button"
          >
            <span
              class="btn__inner"
            >
              Create New Batch
            </span>
          </button>
        </a>
        <h4
          class="fieldset__legend"
        >
          Filter Batches
        </h4>
        <p
          class="checkboxes__label"
        >
          Batch Status
        </p>
        <span
          class="radios__items"
        >
          <p
            class="radios__item"
            style="margin-right: .5rem; min-width: 15rem;"
          >
            <span
              class="radio"
            >
              <input
                checked=""
                class="radio__input js-radio"
                id="live"
                name="Live"
                type="radio"
                value="live"
              />
              <label
                class="radio__label "
                for="live"
              >
                Live
              </label>
            </span>
          </p>
          <p
            class="radios__item"
            style="margin-right: .5rem; min-width: 15rem;"
          >
            <span
              class="radio"
            >
              <input
                class="radio__input js-radio"
                id="completed"
                name="Completed"
                type="radio"
                value="completed"
              />
              <label
                class="radio__label "
                for="completed"
              >
                Completed
              </label>
            </span>
          </p>
          <br />
        </span>
        <p
          class="checkboxes__label"
        >
          Batch Type
        </p>
        <span
          class="checkboxes__items"
        >
          <p
            class="checkboxes__item"
            style="margin-right: .5rem; min-width: 15rem;"
          >
            <span
              class="checkbox "
            >
              <input
                checked=""
                class="checkbox__input js-checkbox "
                id="Monthly"
                name="Monthly"
                type="checkbox"
                value="Monthly"
              />
              <label
                class="checkbox__label "
                for="Monthly"
              >
                Monthly
              </label>
            </span>
          </p>
          <p
            class="checkboxes__item"
            style="margin-right: .5rem; min-width: 15rem;"
          >
            <span
              class="checkbox "
            >
              <input
                checked=""
                class="checkbox__input js-checkbox "
                id="Quarterly"
                name="Quarterly"
                type="checkbox"
                value="Quarterly"
              />
              <label
                class="checkbox__label "
                for="Quarterly"
              >
                Quarterly
              </label>
            </span>
          </p>
          <p
            class="checkboxes__item"
            style="margin-right: .5rem; min-width: 15rem;"
          >
            <span
              class="checkbox "
            >
              <input
                checked=""
                class="checkbox__input js-checkbox "
                id="Annually"
                name="Annually"
                type="checkbox"
                value="Annually"
              />
              <label
                class="checkbox__label "
                for="Annually"
              >
                Annually
              </label>
            </span>
          </p>
        </span>
      </fieldset>
      <table
        class="table table--sortable"
        data-aria-asc="ascending"
        data-aria-desc="descending"
        data-aria-sort="Sort by"
        id="basic-table"
      >
        <thead
          class="table__head"
        >
          <tr
            class="table__row"
          >
            <th
              aria-sort="none"
              class="table__header "
              scope="col"
            />
            <th
              class="table__header "
              scope="col"
            >
              BatchID
            </th>
            <th
              class="table__header "
              scope="col"
            >
              Batch Type
            </th>
            <th
              aria-sort="none"
              class="table__header "
              scope="col"
            >
              <button
                aria-label="Sort by Period"
                class="table__sort-button"
                data-index="2"
                type="button"
              >
                Period
              </button>
            </th>
            <th
              aria-sort="none"
              class="table__header "
              scope="col"
            >
              <button
                aria-label="Sort by Year"
                class="table__sort-button"
                data-index="3"
                type="button"
              >
                Year
              </button>
            </th>
            <th
              aria-sort="none"
              class="table__header "
              scope="col"
            >
              <button
                aria-label="Sort by Status"
                class="table__sort-button"
                data-index="4"
                type="button"
              >
                Status
              </button>
            </th>
          </tr>
        </thead>
        <tbody
          class="table__body expandedRowEnabled "
        >
          <tr>
            <td
              class="table__cell "
              colspan="6"
            >
              <div
                class="panel panel--info panel--simple "
              >
                <div
                  class="panel__body"
                >
                  <p>
                    No Batches matching this criteria
                  </p>
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
